File: boj/not_passed/17840.c
#include <stdio.h>
#include <string.h>
int ret[2][2]
int matsq(int arr[][2], int brr[][2])
{
    memset(ret, 0, sizeof(ret))
    int i, j, k;
    for (i = 0; i < 2; i++)
        for (j = 0; j < 2; j++)
            for (k = 0; k < 2; k ++)
                ret[i][j] = (ret[i][j] + arr[i][k] * brr[k][j]) % m          
}
int fib(int n, int m)
{
    int fb[2][2] = {{1, 1}, {1, 0}}, ans = {{1, 0}, {0, 1}}
}

File: boj/not_passed/17840.py
import sys
def mat(a, b, m):
	ret = [[0]*2 for i in range(2)]
	for i in range(2):
		for j in range(2):
			for k in range(2):
				ret[i][j] = (ret[i][j]+a[i][k]*b[k][j])%m
	return ret
def fib(n, m):
    l = [[1,1],[1,0]]
    ans = [[1,0],[0,1]]
    while(n):
        if n%2:
            ans = mat(ans, l, m)
        l = mat(l, l, m)
        n//=2
    return ans[1][0]        
q, m = map(int, input().split())
for i in range(q):
    a = int(sys.stdin.readline())
    print(fib(a, m)%m)